fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1584,16 +1573,49 @@ Future<std::vector<std::pair<TenantName, TenantMapEntry>>> listTenants(
     TenantName begin,
     TenantName end,
     int limit,
+    int offset = 0,
     std::vector<TenantState> filters = std::vector<TenantState>()) {
 	state Reference<typename DB::TransactionT> tr = db->createTransaction();
 
 	loop {
 		try {
 			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
 			tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
-			std::vector<std::pair<TenantName, TenantMapEntry>> tenants =
-			    wait(listTenantsTransaction(tr, begin, end, limit, filters));
-			return tenants;
+			if (offset == 0 && filters.empty()) {
+				std::vector<std::pair<TenantName, TenantMapEntry>> tenants =
+				    wait(listTenantsTransaction(tr, begin, end, limit));
+				return tenants;
+			}
+			tr->setOption(FDBTransactionOptions::RAW_ACCESS);
+
+			state KeyBackedRangeResult<std::pair<TenantName, TenantMapEntry>> results =
+			    wait(ManagementClusterMetadata::tenantMetadata().tenantMap.getRange(tr, begin, end, limit));
+			state std::vector<std::pair<TenantName, TenantMapEntry>> filterResults;
+			state int count = 0;
+			loop {
+				for (auto pair : results.results) {
+					if (filters.empty() || std::count(filters.begin(), filters.end(), pair.second.tenantState)) {
+						++count;
+						if (count > offset) {
+							filterResults.push_back(pair);
+							if (count - offset == limit) {
+								ASSERT(count - offset == filterResults.size());
+								return filterResults;
+							}
+						}
+					}
+				}
+				if (!results.more) {
+					return filterResults;
+				}
+				if (results.readThrough.present()) {
+					begin = results.readThrough.get();
```

Comment:
I'm not sure we can use `readThrough` this way here, because if I understand correctly it is a raw key rather than an object matching the type of the key backed map. It is probably good to have this ability though, so we may need another way to determine where to start.

Reply:
Are you saying we should continue trying to use `readThrough`, but manipulate the raw key, or just fallback to the `else` case at all times?

Reply:
I'm saying we probably still want to have a way to pick up where we actually left off, but I don't think this works because the result returned is not of the same type as the items being queried.